doc: add comment about highWaterMark limit (#33432)
nodejs/node: doc/api/stream.md

is to limit the buffering of data to acceptable levels such that sources and
destinations of differing speeds will not overwhelm the available memory.

> Watermarks only determine the amount of data a stream buffers before it stops
> asking for more data (via calling `readable._read()`). Watermarks do not
> enforce a strict memory limitation in general. Specific stream
> implementations may choose to enforce stricter limits but doing so is
> optional. Specifically, `highWaterMark` is a threshold, not a limit.

Reviewer: - The word "watermark" is not used anywhere else in the document
- Open with the key point (`highWaterMark` is a threshold) then proceed to explain it
- "May choose ... doing so is optional" is double
- "Via calling `readable._read()`" is a detail not needed for this explanation IMO

```suggestion
The `highWaterMark` option is a threshold, not a limit: it dictates the amount 
of data that a stream buffers before it stops asking for more data. Specific 
stream implementations may choose to enforce stricter limits.
```

Author: Excellent suggestion, thanks.
